nomos-da/network/core/src/protocols/replication/behaviour_v2.rs

```diff
@@ -0,0 +1,677 @@
+                error: std::io::Error::new(error.kind(), error.to_string()),
+            },
+            Self::OpenStream { peer_id, error } => Self::OpenStream {
+                peer_id: *peer_id,
+                error: match error {
+                    OpenStreamError::UnsupportedProtocol(protocol) => {
+                        OpenStreamError::UnsupportedProtocol(protocol.clone())
+                    }
+                    OpenStreamError::Io(error) => {
+                        OpenStreamError::Io(std::io::Error::new(error.kind(), error.to_string()))
+                    }
+                    err => OpenStreamError::Io(std::io::Error::new(
+                        std::io::ErrorKind::Other,
+                        err.to_string(),
+                    )),
+                },
+            },
+        }
+    }
+}
+
+/// Nomos DA `BroadcastEvents` to be bubbled up to logic layers
+#[derive(Debug)]
+pub enum ReplicationEvent {
+    IncomingMessage {
+        peer_id: PeerId,
+        message: Box<DaMessage>,
+    },
+    ReplicationError {
+        error: ReplicationError,
+    },
+}
+
+impl From<ReplicationError> for ReplicationEvent {
+    fn from(error: ReplicationError) -> Self {
+        Self::ReplicationError { error }
+    }
+}
+
+impl ReplicationEvent {
+    #[must_use]
+    pub fn blob_size(&self) -> Option<usize> {
+        match self {
+            Self::IncomingMessage { message, .. } => Some(message.blob.data.column_len()),
+            Self::ReplicationError { .. } => None,
+        }
+    }
+}
+
+type IncomingTask = BoxFuture<'static, Result<(PeerId, DaMessage, Stream), ReplicationError>>;
+type OutboundTask = BoxFuture<'static, Result<(PeerId, Stream), ReplicationError>>;
+
+enum StreamState {
+    Idle(Stream),
+    Busy,
+}
+
+impl StreamState {
+    pub fn take(&mut self) -> Self {
+        let mut ret = Self::Busy;
+        std::mem::swap(self, &mut ret);
+        ret
+    }
+}
+
+/// Holds pending outbound messages for each peer and manages which message
+/// should be scheduled next for writing to an outbound stream.
+#[derive(Default)]
+struct PendingOutbound {
+    /// Pending outbound message queues for each peer
+    messages: IndexMap<PeerId, VecDeque<DaMessage>>,
+    /// The last peer whose pending message was scheduled for sending
+    last_scheduled: Option<PeerId>,
+}
+
+impl PendingOutbound {
+    /// Appends a message to the peer's pending queue. Lazily creates the queue
+    /// for the peer upon the first call.
+    fn enqueue_message(&mut self, peer_id: PeerId, message: DaMessage) {
+        self.messages.entry(peer_id).or_default().push_back(message);
+    }
+
+    /// ### Panics
+    ///
+    /// [`Self::enqueue_message`] must be called at least once for the peer
+    /// before calling this method or it will panic.
+    fn dequeue_message(&mut self, peer_id: &PeerId) -> Option<DaMessage> {
+        self.messages
+            .get_mut(peer_id)
+            .expect("Peer is in the map")
```

Comment:
Couldn't the `expect` here be avoided an `None` returned instead?

Reply:
Indeed. And probably just logged that the peer is missing in that case.